File: src/SnapToTable.Domain/Common/PagedResult.cs
namespace SnapToTable.Domain.Common;

public record PagedResult<T>
{
    public IReadOnlyList<T> Items { get; }
    public long TotalCount { get; }
    public int Page { get; }
    public int PageSize { get; }
    public int TotalPages => (int)Math.Ceiling(TotalCount / (double)PageSize);
    public PagedResult(IReadOnlyList<T> items, long totalCount, int page, int pageSize)
    {
        Items = items;
        TotalCount = totalCount;
        Page = page;
        PageSize = pageSize;
    }
}

File: client/snap-to-table/features/recipes/types/pagedResult.ts
export default class PagedResult<T> {
    public readonly items: readonly T[];
    public readonly totalCount: number;
    public readonly page: number;
    public readonly pageSize: number;

    constructor(items: readonly T[], totalCount: number, page: number, pageSize: number) {
        this.items = items;
        this.totalCount = totalCount;
        this.page = page;
        this.pageSize = pageSize;
    }

    public get totalPages(): number {
        return Math.ceil(this.totalCount / this.pageSize);
    }

    public get hasNextPage(): boolean {
        return this.page < this.totalPages;
    }

    public get hasPreviousPage(): boolean {
        return this.page > 1;
    }
}